src/test/java/com/microsoft/sqlserver/jdbc/ActivityIDTest.java

```diff
@@ -0,0 +1,179 @@
+                                }
+                            } catch (SQLException e) {
+                                fail(e.toString());
+                            }
+                            latchPool.countDown();
+                        }
+                    });
+                }
+                try {
+                    latchPool.await();
+                } catch (InterruptedException e) {
+                    fail(e.toString());
+                } finally {
+                    if (null != ds) {
+                        es.shutdown();
+                        ds.close();
+                    }
+                }
+                latchPoolOuterThread.countDown();
+            }
+        });
+        t.run();
+        latchPoolOuterThread.await();
+        
+        try {
+            try (Connection con = getConnection(); Statement stmt = con.createStatement()) {
+                stmt.execute("SELECT @@VERSION AS 'SQL Server Version'");
+            }
+        } catch (SQLException e) {
+            fail(e.toString());
+        }
+        
+        try {
+            assertEquals(0, ActivityCorrelator.getActivityIdTlsMap().size());
+        } catch (ComparisonFailure e) {
+            Set<Thread> threadSet = Thread.getAllStackTraces().keySet();
+            logger.fine("List of threads alive:");
+            for (Thread thread: threadSet) {
+                logger.fine(String.valueOf(thread.getId()));
+            }
+            logger.fine("List of entries in the ActivityID map:");
+            logger.fine(ActivityCorrelator.getActivityIdTlsMap().toString());
+            throw new Exception(e);
+        }
+    }
+    
+    @Test
+    public void testActivityIDPooledConnection() throws Exception {
+        int poolsize = 10;
+        int numPooledExecution = 200;
+
+        PooledConnection pooledCon = ((SQLServerConnectionPoolDataSource) dsPool).getPooledConnection();
+        ExecutorService es = Executors.newFixedThreadPool(poolsize);
+        try {
+            CountDownLatch latchPool = new CountDownLatch(numPooledExecution);
+            es.execute(new Runnable() {
+                public void run() {
+                    for (int i = 0; i < numPooledExecution; i++) {
+                        try (Connection con = pooledCon.getConnection(); Statement stmt = con.createStatement()) {
+                            stmt.execute("SELECT @@VERSION AS 'SQL Server Version'");
+                        } catch (SQLException e) {
+                            fail(e.toString());
+                        }
+                        latchPool.countDown();
+                    }
+                }
+            });
+            latchPool.await();
+        } finally {
+            es.shutdown();
+            pooledCon.close();
+        }
+        assertEquals(0, ActivityCorrelator.getActivityIdTlsMap().size());
+    }
+    
+    @BeforeEach
+    @AfterEach
+    public void clearActivityId() {
+        ActivityCorrelator.clear();

```

Comment:
All of these tests assume ActivityCorrelator size should be 0 at the end of the tests. If this is necessary, doesn't that imply there's something inherently wrong with the tests? Also I don't think it needs to be both Before and After.

Reply:
The problem is only with HikariCP test - but I think this part can be removed since we figured out an alternative to clear activity ID for Hikari's thread - @peterbae can we run test without this method?

Reply:
I agree it's redundant after the changes we made. I've removed this part.